src/Types/LazyTypeRegistry.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Bundle\MongoDBBundle\Types;
+
+use Doctrine\ODM\MongoDB\Types\Type;
+use Doctrine\ODM\MongoDB\Types\TypeRegistry;
+use InvalidArgumentException;
+use Symfony\Component\DependencyInjection\ServiceLocator;
+use Symfony\Contracts\Service\ServiceProviderInterface;
+
+use function sprintf;
+
+/**
+ * Field type registry that lazy-loads types from a service locator.
+ */
+class LazyTypeRegistry extends TypeRegistry
+{
+    /** @param ServiceLocator<Type> $locator */
+    public function __construct(private ServiceProviderInterface $locator)
+    {
+    }
+
+    public function has(string $name): bool
+    {
+        return $this->locator->has($name) || parent::has($name);
```

Comment:
I assume there's no significant performance difference between the two (as both are essentially maps connecting a string to an object), so the order of operations does not matter here.

Reply:
Indeed, default types are a lot more used than the custom ones. 
```suggestion
        return parent::has($name) || $this->locator->has($name);
```